### PR TITLE
Data generalization

### PR DESCRIPTION
**Pull request by anna-boser**
Here, I make the stagg package able to be used with data from sources other than ERA5 by:
- Making it functional with any raster with layer names following the format: any non-digit characters followed by a date compatible with lubridate::as_datetime()
- Making it functional for rasters without adequate layer names by adding the optional `start_date` and `time_interval` arguments in staggregate_*()
- Note that if the user wants the daily_agg but the data is not hourly, they MUST supply the `time_interval` argument in staggregate_*() **even if the layer names are compatible with stagg**, otherwise the default of 1 hour will be used. Note that in theory we could fix this in the future if we did the daily agg based on layer names as well instead of based on the number of timesteps in a day which we infer from `time_interval`. I think we didn't do it in this way originally for computational efficiency.

### CODE REVIEW

File: R/staggregate.R [hunk 1]
```diff
@@ -154,6 +172,28 @@ daily_aggregation <- function(data, overlay_weights, daily_agg){
   return(list(clim_daily, layer_names))
 }
 
+# Function to infer date-times for raster layers based on a time interval
+infer_layer_datetimes <- function(raster_stack, start_date, time_interval) {
```
Comment (saraorofino) [inline R/staggregate.R:176]: What happens if this is the default NA? Does this part skip if no argument is applied to start_date?

File: R/staggregate.R [hunk 2]
```diff
@@ -260,6 +299,16 @@ polygon_aggregation <- function(clim_dt, weights_dt, list_names, time_agg){
...
+#' @param start_date the date (and time, if applicable) of the first layer in
+#'  the raster. To be input in a format compatible with
+#'  lubridate::as_datetime(), e.g. `"1991-10-29"` or `"1991-10-29 00:00:00"`.
+#'  The default is `NA` since the rasters usually already contain temporal
+#'  information in the layer names and they do not need to be manually supplied.
+#' @param time_interval the time interval between layers in the raster to be
+#'  aggregated. To be input in a format compatible with seq(), e.g.
+#'  `'1 day'` or `'3 months'`. The default is `'1 hour'` and this argument is
+#'  required if daily_agg is not `'none'` or if the layer name format is not
+#'  compatible with stagg.
```
Comment (saraorofino) [inline R/staggregate.R:311]: Maybe we can update just this last part of "or if layer name format is not compatible with stagg" to be "or if the start_date argument is not NA" (as long as that is true)

File: R/staggregate.R [hunk 3]
```diff
@@ -269,14 +318,21 @@ polygon_aggregation <- function(clim_dt, weights_dt, list_names, time_agg){
 #'   daily_agg = "average", # Average hourly values to produce daily values
 #'                          # before transformation
 #'   time_agg = "month", # Sum the transformed daily values across months
+#'   start_date = "2020-01-01 00:00:00", # The start date of the supplied data, only required if the layer name format is not compatible with stagg
+#'   time_interval = "1 hour", # The temporal interval of the supplied data, required if daily_agg is not "none" or if the layer name format is not compatible with stagg
```
Comment (saraorofino) [inline R/staggregate.R:322]: Same comment here on the guidance for time_interval.  Needs to be supplied when daily_agg is not "none" or start_date is not "NA"

File: R/staggregate.R [hunk 4]
```diff
@@ -364,6 +420,16 @@ staggregate_polynomial <- function(data, overlay_weights, daily_agg, time_agg =
...
+#' @param start_date the date (and time, if applicable) of the first layer in
+#'  the raster. To be input in a format compatible with
+#'  lubridate::as_datetime(), e.g. `"1991-10-29"` or `"1991-10-29 00:00:00"`.
+#'  The default is `NA` since the rasters usually already contain temporal
+#'  information in the layer names and they do not need to be manually supplied.
+#' @param time_interval the time interval between layers in the raster to be
+#'  aggregated. To be input in a format compatible with seq(), e.g.
+#'  `'1 day'` or `'3 months'`. The default is `'1 hour'` and this argument is
+#'  required if daily_agg is not `'none'` or if the layer name format is not
+#'  compatible with stagg.
```
Comment (saraorofino) [inline R/staggregate.R:432]: Same here on the time_interval guidance (just want to make sure we don't miss any)

File: R/staggregate.R [hunk 5]
```diff
@@ -373,13 +439,21 @@ staggregate_polynomial <- function(data, overlay_weights, daily_agg, time_agg =
 #' daily_agg = "average", # Average hourly values to produce daily values before
 #'                        # transformation
 #' time_agg = "month", # Sum the transformed daily values across months
+#' start_date = "2020-01-01 00:00:00", # The start date of the supplied data, only required if the layer name format is not compatible with stagg
+#' time_interval = "1 hour", # The temporal interval of the supplied data, required if daily_agg is not "none" or if the layer name format is not compatible with stagg
```
Comment (saraorofino) [inline R/staggregate.R:443]: And here

File: R/staggregate.R [hunk 6]
```diff
@@ -495,6 +569,16 @@ staggregate_spline <- function(data, overlay_weights, daily_agg, time_agg = "mon
...
+#' @param start_date the date (and time, if applicable) of the first layer in
+#'  the raster. To be input in a format compatible with
+#'  lubridate::as_datetime(), e.g. `"1991-10-29"` or `"1991-10-29 00:00:00"`.
+#'  The default is `NA` since the rasters usually already contain temporal
+#'  information in the layer names and they do not need to be manually supplied.
+#' @param time_interval the time interval between layers in the raster to be
+#'  aggregated. To be input in a format compatible with seq(), e.g.
+#'  `'1 day'` or `'3 months'`. The default is `'1 hour'` and this argument is
+#'  required if daily_agg is not `'none'` or if the layer name format is not
+#'  compatible with stagg.
```
Comment (saraorofino) [inline R/staggregate.R:581]: And here

File: R/staggregate.R [hunk 7]
```diff
@@ -504,14 +588,22 @@ staggregate_spline <- function(data, overlay_weights, daily_agg, time_agg = "mon
 #'   daily_agg = "average", # Average hourly values to produce daily values
 #'                          # before transformation
 #'   time_agg = "month", # Sum the transformed daily values across months
+#'   start_date = "2020-01-01 00:00:00", # The start date of the supplied data, only required if the layer name format is not compatible with stagg
+#'   time_interval = "1 hour", # The temporal interval of the supplied data, required if daily_agg is not "none" or if the layer name format is not compatible with stagg
```
Comment (saraorofino) [inline R/staggregate.R:592]: And here

File: R/staggregate.R [hunk 8]
```diff
@@ -624,23 +716,40 @@ staggregate_bin <- function(data, overlay_weights, daily_agg, time_agg = "month"
...
 #'   `'month'`, or `'year'`
+#' @param start_date the date (and time, if applicable) of the first layer in
+#'  the raster. To be input in a format compatible with
+#'  lubridate::as_datetime(), e.g. `"1991-10-29"` or `"1991-10-29 00:00:00"`.
+#'  The default is `NA` since the rasters usually already contain temporal
+#'  information in the layer names and they do not need to be manually supplied.
+#' @param time_interval the time interval between layers in the raster to be
+#'  aggregated. To be input in a format compatible with seq(), e.g.
+#'  `'1 day'` or `'3 months'`. The default is `'1 hour'` and this argument is
+#'  required if the layer name format is not compatible with stagg.
```
Comment (saraorofino) [inline R/staggregate.R:727]: Here

File: R/staggregate.R [hunk 9]
```diff
@@ -624,23 +716,40 @@ staggregate_bin <- function(data, overlay_weights, daily_agg, time_agg = "month"
...
 #' @param thresholds A vector of temperature thresholds critical to a crop
 #'
 #' @examples
 #' degree_days_output <- staggregate_degree_days(
 #'   data = temp_kansas_jan_2020_era5, # Climate data to transform and aggregate
 #'   overlay_weights = overlay_weights_kansas, # Output from overlay_weights()
 #'   time_agg = "month", # Sum the transformed daily values across months
+#'   start_date = "2020-01-01 00:00:00", # The start date of the supplied data, only required if the layer name format is not compatible with stagg
+#'   time_interval = "1 hour", # The temporal interval of the supplied data, only required if the layer name format is not compatible with stagg
```
Comment (saraorofino) [inline R/staggregate.R:736]: Here

**Review by saraorofino — CHANGES_REQUESTED** [review]
Just want to confirm the requirements for the time_interval parameter (e.g., if start_date is not NA then is this parameter required?) and update that across all the staggregate functions as needed

File: R/staggregate.R (hunk 1 above)
Comment (anna-boser) [reply R/staggregate.R:176]: You only call infer_layer_datetimes if the start_date is not NA -- if it's the default NA you use the existing layer names

**Review by saraorofino — APPROVED** [review]
All feedback addressed and changes look good

**Pull request merged**